control-plane/connect-inject/controllers/pod/pod_controller.go

```diff
@@ -432,10 +445,266 @@ func (r *Controller) writeHealthStatus(ctx context.Context, pod corev1.Pod) erro
+			case "peer":
+				// TODO: uncomment and remove error when peers supported
+				//peer = strings.TrimSpace(pieces[6])
+				return &pbmesh.Upstream{}, fmt.Errorf("upstream currently does not support peers: %s", rawUpstream)
+			case "ap":
+				partition = strings.TrimSpace(pieces[6])
+			case "dc":
+				// TODO: uncomment and remove error when datacenters are supported
+				//datacenter = strings.TrimSpace(pieces[6])
+				return &pbmesh.Upstream{}, fmt.Errorf("upstream currently does not support datacenters: %s", rawUpstream)
+			default:
+				return &pbmesh.Upstream{}, fmt.Errorf("upstream structured incorrectly: %s", rawUpstream)
+			}
+			fallthrough
+		case 6:
+			if strings.TrimSpace(pieces[5]) == "ns" {
+				namespace = strings.TrimSpace(pieces[4])
+			} else {
+				return &pbmesh.Upstream{}, fmt.Errorf("upstream structured incorrectly: %s", rawUpstream)
+			}
+			fallthrough
+		case 4:
+			if strings.TrimSpace(pieces[3]) == "svc" {
+				svcName = strings.TrimSpace(pieces[2])
+			} else {
+				return &pbmesh.Upstream{}, fmt.Errorf("upstream structured incorrectly: %s", rawUpstream)
+			}
+			if strings.TrimSpace(pieces[1]) == "port" {
+				portName = strings.TrimSpace(pieces[0])
+			} else {
+				return &pbmesh.Upstream{}, fmt.Errorf("upstream structured incorrectly: %s", rawUpstream)
+			}
+		default:
+			return &pbmesh.Upstream{}, fmt.Errorf("upstream structured incorrectly: %s", rawUpstream)
+		}
+	} else {
+		switch len(pieces) {
+		case 6:
+			end := strings.TrimSpace(pieces[5])
+			switch end {
+			case "peer":
+				// TODO: uncomment and remove error when peers supported
+				//peer = strings.TrimSpace(pieces[4])
+				return &pbmesh.Upstream{}, fmt.Errorf("upstream currently does not support peers: %s", rawUpstream)
+			case "dc":
+				// TODO: uncomment and remove error when datacenter supported
+				//datacenter = strings.TrimSpace(pieces[4])
+				return &pbmesh.Upstream{}, fmt.Errorf("upstream currently does not support datacenters: %s", rawUpstream)
+			default:
+				return &pbmesh.Upstream{}, fmt.Errorf("upstream structured incorrectly: %s", rawUpstream)
+			}
+			// TODO: uncomment and remove error when datacenter and/or peers supported
+			//fallthrough
+		case 4:
+			if strings.TrimSpace(pieces[3]) == "svc" {
+				svcName = strings.TrimSpace(pieces[2])
+			} else {
+				return &pbmesh.Upstream{}, fmt.Errorf("upstream structured incorrectly: %s", rawUpstream)
+			}
+			if strings.TrimSpace(pieces[1]) == "port" {
+				portName = strings.TrimSpace(pieces[0])
+			} else {
+				return &pbmesh.Upstream{}, fmt.Errorf("upstream structured incorrectly: %s", rawUpstream)
+			}
+		default:
+			return &pbmesh.Upstream{}, fmt.Errorf("upstream structured incorrectly: %s", rawUpstream)
+		}
+	}
+
+	var upstream pbmesh.Upstream
+	if port > 0 {
+		upstream = pbmesh.Upstream{
+			DestinationRef: &pbresource.Reference{
+				Tenancy: &pbresource.Tenancy{
+					Partition: partition,
+					Namespace: namespace,
+					PeerName:  peer,
```

Comment:
Since Consul V2 is weird about defaults, we'll want to check if these are empty and set them explicitly to constant values.